File: Content/Script/UnLuaPerformanceTestProxy.lua
require "UnLua"

local UnLuaPerformanceTestProxy = Class()

function UnLuaPerformanceTestProxy:TestObjectDelegate(myints,mybool)
	for i = 1, myints:Length() do
		print("myintarray:" .. myints:Get(i))
	end
	print(mybool)
end

function UnLuaPerformanceTestProxy:TestObjectDelegate2(myvector,mybool)
	print(myvector.X)
	print(myvector.Y)
	print(myvector.Z)
	print(mybool)
end

function UnLuaPerformanceTestProxy:TestDelegatePrint(myint,myfloat)
	print("myint:" .. myint)
	print("myfloat:" .. myfloat)
end

function UnLuaPerformanceTestProxy:NativeEvent()
	print("this is lua native event")
	--self.Overridden.NativeEvent(self)--crash
end


function UnLuaPerformanceTestProxy:ImplementationEvent()
	print("this is lua Implementation event")
	self.Overridden.ImplementationEvent(self)
end

function UnLuaPerformanceTestProxy:ReceiveBeginPlay()
	local N = 1000000
	local Multiplier = 1000000000.0 / N
	local RawObject = self.Object

	-- warm up
	self:NOP()


	self:AddToMultiDelegate({self,UnLuaPerformanceTestProxy.TestDelegatePrint})
	self:BroadcastMultiDelegate(100,123.123)
	self:ClearMultiDelegate()

	-- local delegate = self:GetDelegate()
	-- delegate:Execute(345,345.234)
	-- delegate:Execute(111,3456.234)
	-- delegate:Execute(222,3457.234)
	-- delegate:Unbind()

	-- local a = self:GetAAA()
	-- print(a)
	-- self:printA()
	-- a = a + 100
	-- print(a)
	-- self:printA()

	-- -- warm up
	-- for i=1, 1 do
	-- 	self:NOP()
	-- end
	
	RawObject:SetDelegate({self,UnLuaPerformanceTestProxy.TestDelegatePrint},999,999.7765)
	RawObject.TestDelegate:Unbind()


	RawObject.TestDelegate:Bind(self,UnLuaPerformanceTestProxy.TestDelegatePrint)
	RawObject.TestDelegate:Execute(345,345.234)
	RawObject.TestDelegate:Execute(111,3456.234)
	RawObject.TestDelegate:Execute(222,3457.234)
	RawObject.TestDelegate:Unbind()

	local TestStrings = UE4.TArray(UE4.FString)
	for i=1, 10 do
		local str = "str" .. i
		TestStrings:Add(str)
	end
	StartTime = Seconds()
	for i=1, 10 do
		self:ServerUpdateStrings(TestStrings)
	end
	EndTime = Seconds()

	local TestInts = UE4.TArray(int32)
	for i=1, 10 do
		TestInts:Add(i)
	end
	StartTime = Seconds()
	for i=1, 10 do
		self:ServerUpdateInts(TestInts)
	end
	EndTime = Seconds()

	local Objects = UE4.TArray(UE4.UClassObject)
	StartTime = Seconds()
	for i=1, 10 do
		self:GetObjects(Objects)
	end
	print("object lenth:" .. Objects:Length())
	for i=1,Objects:Length() do
		local obj = Objects:Get(i)

		local COM = UE4.FVector(1.1, 1.2, 1.3)
		local strcom = tostring(COM)--ok
		print(strcom)

		local COM2 = UE4.FVector()
		local strcom2 = tostring(COM2)
		--local strcom = tostring(UE4.FVector)--crash
		print(strcom2)

		print("===01")
		obj.mymultidelegate:Add(self,UnLuaPerformanceTestProxy.TestObjectDelegate2)
		print("===02")
		obj.mymultidelegate:Broadcast(COM,true)
		print("===03")
		obj.mymultidelegate:Broadcast(COM,false)--crash
		print("===04")
		obj.mymultidelegate:Broadcast(COM,false)
		print("===05")
		obj.mymultidelegate:Broadcast(COM,false)
		print("===06")
		obj.mymultidelegate:Broadcast(COM,true)
		print("===07")
		obj.mymultidelegate:Broadcast(COM,false)
		print("===08")
		obj.mymultidelegate:Broadcast(COM,false)
		print("===09")
		obj.mymultidelegate:Broadcast(COM,false)
		print("===00")
		obj.mymultidelegate:Clear()

		for i=1,10 do
			print("===1")
			obj.mydelegate:Bind(self,UnLuaPerformanceTestProxy.TestObjectDelegate)
			print("===2")
			obj.mydelegate:Execute(TestInts,true)--循环第二次执行到这里的时候崩溃
			print("===3")
			obj.mydelegate:Execute(TestInts,false)
			print("===4")
			obj.mydelegate:Unbind()
		end


		print("===11")
		obj.mydelegate2:Bind(self,UnLuaPerformanceTestProxy.TestObjectDelegate2)
		print("===22")
		obj.mydelegate2:Execute(COM,true)
		print("===33")
		obj.mydelegate2:Execute(COM,false)
		print("===44")
		obj.mydelegate2:Unbind()

		print(Objects:Get(i).type)
	end
	EndTime = Seconds()
	
	-- local StartTime = Seconds()
	-- for i=1, N do
	-- 	local MeshID = RawObject.MeshID
	-- end
	-- local EndTime = Seconds()
	-- local Message = "read int32 ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	RawObject.MeshID = i
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" ..  "write int32 ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local MeshName = RawObject.MeshName
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" ..  "read FString ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	RawObject.MeshName = "9527"
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" ..  "write FString ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local COM = RawObject.COM
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" ..  "read FVector ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- local COM = UE4.FVector(1.0, 1.0, 1.0)
	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	RawObject.COM = COM
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" ..  "write FVector ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local Positions = RawObject.Positions
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" ..  "read TArray<FVector> ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- local PredictedPositions = UE4.TArray(UE4.FVector)
	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	RawObject.PredictedPositions = PredictedPositions
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" ..  "write TArray<FVector> ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:NOP()
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" ..  "void NOP() ; " .. tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:Simulate(0.0167)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "void Simulate(float) ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local MeshID = self:GetMeshID()
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "int32 GetMeshID() const ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local MeshName = self:GetMeshName()
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "const FString& GetMeshName() const ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:GetCOM(COM)
	-- 	--local COMCopy = self:GetCOM(COM)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "const FVector& GetCOM() const ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local NewMeshID = self:UpdateMeshID(1024)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "int32 UpdateMeshID(int32) ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local NewMeshName = self:UpdateMeshName("1024")
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "FString UpdateMeshName(const FString&) ; "..tostring((EndTime - StartTime) * Multiplier)

	-- local Origin = UE4.FVector(1.0, 1.0, 1.0)
	-- local Direction = UE4.FVector(1.0, 0.0, 0.0)
	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local bHit = self:Raycast(Origin, Direction)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "bool Raycast(const FVector&, const FVector&) const ; "..tostring((EndTime - StartTime) * Multiplier)

	-- local Indices = UE4.TArray(0)
	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:GetIndices(Indices)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "void GetIndices(TArray<int32>&) const ; "..tostring((EndTime - StartTime) * Multiplier)

	-- for i=1, 1024 do
	-- 	Indices:Add(i)
	-- end
	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:GetIndices(Indices)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "void GetIndices(TArray<int32>&) const with 1024 items ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:UpdateIndices(Indices)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "void UpdateIndices(const TArray<int32>&) ; "..tostring((EndTime - StartTime) * Multiplier)

	-- local Positions = UE4.TArray(UE4.FVector)
	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:GetPositions(Positions)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "void GetPositions(TArray<FVector>&) const ; "..tostring((EndTime - StartTime) * Multiplier)

	-- for i=1, 1024 do
	-- 	Positions:Add(UE4.FVector(i, i, i))
	-- end
	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:GetPositions(Positions)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "void GetPositions(TArray<FVector>&) const with 1024 items ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:UpdatePositions(Positions)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "void UpdatePositions(const TArray<FVector>&) ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	self:GetPredictedPositions(PredictedPositions)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "const TArray<FVector>& GetPredictedPositions() const ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local ID, Name, bResult = self:GetMeshInfo(0, "", COM, Indices, Positions, PredictedPositions)
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "bool GetMeshInfo(int32&, FString&, FVector&, TArray<int32>&, TArray<FVector>&, TArray<FVector>&) const ; "..tostring((EndTime - StartTime) * Multiplier)

	-- StartTime = Seconds()
	-- for i=1, N do
	-- 	local HitResult = UE4.FHitResult()
	-- end
	-- EndTime = Seconds()
	-- Message = Message .. "\n" .. "FHitResult() ; "..tostring((EndTime - StartTime) * Multiplier)

	-- LogPerformanceData(Message)
end

return UnLuaPerformanceTestProxy

File: Plugins/UnLua/Source/UnLua/Private/Tests/UnLuaPerformanceTestProxy.cpp
// Tencent is pleased to support the open source community by making UnLua available.
// 
// Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the MIT License (the "License"); 
// you may not use this file except in compliance with the License. You may obtain a copy of the License at
//
// http://opensource.org/licenses/MIT
//
// Unless required by applicable law or agreed to in writing, 
// software distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and limitations under the License.

#include "UnLuaPerformanceTestProxy.h"
#include "UnLuaBase.h"


void AUnLuaPerformanceTestProxy::NativeEvent_Implementation()
{
	UE_LOG(LogUnLua, Log, TEXT("NativeEvent:%s"), ANSI_TO_TCHAR(__FUNCTION__));
}

void AUnLuaPerformanceTestProxy::printA()
{
	UE_LOG(LogUnLua, Log, TEXT("printA:%d"), a );
	a++;
}
void AUnLuaPerformanceTestProxy::testBindFunc(int32 myint, float myfloat)
{
	UE_LOG(LogUnLua, Log, TEXT("testBindFunc:%d,%f"), myint, myfloat);
}

void AUnLuaPerformanceTestProxy::NOP()
{

	UClassObject * OBJ = NewObject<UClassObject>();
	OBJ->type = 1;
	Objects.Push(OBJ);
	UClassObject * OBJ2 = NewObject<UClassObject>();
	OBJ2->type = 22;
	Objects.Push(OBJ2);
	UClassObject * OBJ3 = NewObject<UClassObject>();
	OBJ3->type = 333;
	Objects.Push(OBJ3);

	NativeEvent();
	ImplementationEvent();

	TestDelegate.BindUFunction(this, GET_FUNCTION_NAME_CHECKED(AUnLuaPerformanceTestProxy, testBindFunc));
}

void AUnLuaPerformanceTestProxy::Simulate(float DeltaTime)
{
}

int32 AUnLuaPerformanceTestProxy::GetMeshID() const
{
    return MeshID;
}

const FString& AUnLuaPerformanceTestProxy::GetMeshName() const
{
    return MeshName;
}

const FVector& AUnLuaPerformanceTestProxy::GetCOM() const
{
    return COM;
}

int32 AUnLuaPerformanceTestProxy::UpdateMeshID(int32 NewID)
{
    return NewID;
}

FString AUnLuaPerformanceTestProxy::UpdateMeshName(const FString &NewName)
{
    return NewName;
}

bool AUnLuaPerformanceTestProxy::Raycast(const FVector &Origin, const FVector &Direction) const
{
    return true;
}

void AUnLuaPerformanceTestProxy::GetIndices(TArray<int32> &OutIndices) const
{
}

void AUnLuaPerformanceTestProxy::UpdateIndices(const TArray<int32> &NewIndices)
{
}

void AUnLuaPerformanceTestProxy::GetPositions(TArray<FVector> &OutPositions) const
{
}

void AUnLuaPerformanceTestProxy::UpdatePositions(const TArray<FVector> &NewPositions)
{
}

const TArray<FVector>& AUnLuaPerformanceTestProxy::GetPredictedPositions() const
{
    return PredictedPositions;
}

const TArray<UClassObject*>& AUnLuaPerformanceTestProxy::GetObjects() const
{
	return Objects;
}


void AUnLuaPerformanceTestProxy::ServerUpdateStrings_Implementation(const TArray<FString> & NewStrings)
{
	for (int i = 0; i < NewStrings.Num(); i++)
	{
		UE_LOG(LogUnLua, Log, TEXT("function:%s string:%s") , ANSI_TO_TCHAR(__FUNCTION__), *NewStrings[i]);
	}
}

void AUnLuaPerformanceTestProxy::ServerUpdateInts_Implementation(const TArray<int32> & newarray)
{
	for (int i=0;i<newarray.Num();i++)
	{
		FString str = FString::FromInt(newarray[i]);
		UE_LOG(LogUnLua, Log, TEXT("function:%s int:%s"), ANSI_TO_TCHAR(__FUNCTION__), *str);
	}
}

bool AUnLuaPerformanceTestProxy::GetMeshInfo(int32 &OutMeshID, FString &OutMeshName, FVector &OutCOM, TArray<int32> &OutIndices, TArray<FVector> &OutPositions, TArray<FVector> &OutPredictedPositions) const
{
    return true;
}
void AUnLuaPerformanceTestProxy::SetDelegate(FUnLuaPerformanceTestDelegate delegate, int myint, float myfloat)
{
	if (delegate.IsBound())
	{
		TestDelegate = delegate;
		TestDelegate.ExecuteIfBound(myint, myfloat);
	}

}

//#if UE_BUILD_TEST

#include "Misc/DateTime.h"
#include "Misc/FileHelper.h"
#include "UnLuaEx.h"

bool LogPerformanceData(const FString &Message)
{
    FString ProfilingFilePath = FString::Printf(TEXT("%sProfiling/UnLua-Performance-%s.csv"), *FPaths::ConvertRelativePathToFull(FPaths::ProjectSavedDir()), *FDateTime::Now().ToString());
    bool bSuccess = FFileHelper::SaveStringToFile(Message, *ProfilingFilePath);
    return bSuccess;
}

EXPORT_FUNCTION(bool, LogPerformanceData, const FString&)

EXPORT_FUNCTION_EX(Seconds, double, FPlatformTime::Seconds)

//#endif
